File: Math/1071. Greatest Common Divisor of Strings/1071.py
class Solution:
    # runtime could be better. 
    # runtime 47.26%
    # memory 63.35%
    def divisibleString(self, str1: str, str2: str) -> bool:
        '''
        Check if the str1 % str2 == 0. 
    
            Parameters
            ----------
            str1 : str
                The bigger string.
            str2 : str
                The smaller string
    
            Returns
            -------
            bool
                Return True if str1 % str2 == 0. False if str1 % str2 != 0
    
        '''
        if len(str1) % len(str2) != 0:
            return False
        compare = ''
        for i in range(int(len(str1)/len(str2))):
        # len(str1)/len(str2) is float. So i put int() to convert to int
            compare += str2
        return compare == str1
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        # if their lengths are equal, then only need to test if str1 % str2 == 0
        if len(str1) == len(str2):
            if self.divisibleString(str1, str2):
                return str1
            return ""
        # convention: str1 is the larger, str2 is the smaller
        if len(str1)<len(str2):
            tmp = str2
            str2 = str1
            str1 = tmp
            del tmp
        # new str1 is the larger, str2 is the smaller

        # in order to be a gcd, it must be a factor of the smaller string, then a factor of the larger string
        for i in range(len(str2)-1,-1,-1):
            # test if str2[0:i] is a commun factor
            # print(str2[0:i+1])
            # print(i)
            # firstly it must be a factor of the smaller stirng
            if self.divisibleString(str2, str2[0:i+1]):
                # check for the larger string
                if self.divisibleString(str1, str2[0:i+1]):
                    return str2[0:i+1]
        return ""
epsi = Solution()
print(epsi.divisibleString("ABCABC", "ABC"))
epsi.gcdOfStrings("ABCABC", "ABC")

testCases = [("ABCABC", "ABC"), ("ABCABC", "AB"), ("ABCDEF","ABC")]
for testCase in testCases:
    print("Test case: %s %s" % (testCase[0], testCase[1]))
    print("- divisableString : %s" % epsi.divisibleString(testCase[0], testCase[1]))
    print("- gcdOfStrings : %s" % epsi.gcdOfStrings(testCase[0], testCase[1]))
    print("")

File: Math/1071. Greatest Common Divisor of Strings/1071.js
/**
 * check if str1 % str2 == 0
 * @param {string} str1 
 * @param {string} str2 
 * @returns {boolean}
 */
var divisibleString = function(str1, str2){
    if (str1.length % str2.length != 0){
        return false;
    }
    var compare = "";
    var range = Math.floor(str1.length / str2.length);
    for (var i = 0; i < range; i++){
        compare += str2;
    }
    return (compare == str1);
}



/**
 * Returns the greatest common divisor of string
 * @param {string} str1 
 * @param {string} str2 
 * @returns {string}
 */
var gcdOfStrings = function(str1, str2){
    // if their lengths are equal, then only need to test if str1 % str2 == 0
    if (str1.length == str2.length){
        if (divisibleString(str1, str2)){
            return str1;
        }
        return "";
    }
    // convention: str1 is the larger, str2 is the smaller
    if (str1.length == str2.length){
        // Swap values using destructuring assignment
        [str1, str2] = [str2, str1];
    }

    for (var i = str2.length; i>-1; i--){
        // firstly it must be a factor of the smaller stirng
        if (divisibleString(str2, str2.slice(0, i+1))){
            // check for the larger string
            if (divisibleString(str1, str2.slice(0, i+1))){
                return str2.slice(0, i+1);
            }
        }
    }
    return "";
}

var str1 = "hellohello";
var str2 = "helloo";
console.log(gcdOfStrings(str1, str2));

